kube-runtime/src/reflector/store.rs

```diff
@@ -105,39 +108,66 @@ where
                 let key = obj.to_object_ref(self.dyntype.clone());
                 self.store.write().remove(&key);
             }
-            watcher::Event::Restarted(new_objs) => {
+            watcher::Event::RestartedStart => {
+                let mut buffer = self.buffer.write();
+                *buffer = AHashMap::new();
+            }
+            watcher::Event::RestartedPage(new_objs) => {
                 let new_objs = new_objs
                     .iter()
                     .map(|obj| (obj.to_object_ref(self.dyntype.clone()), Arc::new(obj.clone())))
                     .collect::<AHashMap<_, _>>();
-                *self.store.write() = new_objs;
+                self.buffer.write().extend(new_objs);
             }
-        }
+            watcher::Event::RestartedDone => {
+                let mut store = self.store.write();
+
+                // Swap the buffer into the store
+                let mut buffer = self.buffer.write();
+                std::mem::swap(&mut *store, &mut *buffer);
 
-        // Mark as ready after the first event, "releasing" any calls to Store::wait_until_ready()
-        if let Some(ready_tx) = self.ready_tx.take() {
-            ready_tx.init(())
+                // Clear the buffer
+                *buffer = AHashMap::new();
+
+                // Mark as ready after the RestartedDone, "releasing" any calls to Store::wait_until_ready()
+                if let Some(ready_tx) = self.ready_tx.take() {
+                    ready_tx.init(())
+                }
+            }
+            watcher::Event::RestartedApplied(obj) => {
+                let key = obj.to_object_ref(self.dyntype.clone());
+                let obj = Arc::new(obj.clone());
+                self.buffer.write().insert(key, obj);
+            }
+            watcher::Event::RestartedDeleted(obj) => {
+                let key = obj.to_object_ref(self.dyntype.clone());
+                self.buffer.write().remove(&key);
+            }
         }
     }
 
     /// Broadcast an event to any downstream listeners subscribed on the store
     pub(crate) async fn dispatch_event(&mut self, event: &watcher::Event<K>) {
         if let Some(ref mut dispatcher) = self.dispatcher {
             match event {
-                watcher::Event::Applied(obj) => {
+                watcher::Event::Applied(obj) | watcher::Event::RestartedApplied(obj) => {
                     let obj_ref = obj.to_object_ref(self.dyntype.clone());
                     // TODO (matei): should this take a timeout to log when backpressure has
                     // been applied for too long, e.g. 10s
                     dispatcher.broadcast(obj_ref).await;
                 }
 
-                watcher::Event::Restarted(new_objs) => {
+                watcher::Event::RestartedPage(new_objs) => {
```

Comment:
ah. the dispatch interface. I guess we have to answer whether we actually want to propagate pages up to subscribers or not. atm, they are injected into controller interfaces so we could hide pages here, but it does mean that we would have to emit an old-style Restarted event for a subscriber (which does eat away at some of the memory benefits here). maybe @mateiidavid have thoughts here.